XFEM geometric cut switching material system

hugary1995 opened the pull request:
Add SwitchingMaterial, ComboCutUserObject, GeometricCutSubdomainIDAux and tests

close #17041

<!--
INCLUDE THE FOLLOWING IN THE PR DESCRIPTION
- Explain relevant design information for your change.
- Follow the [Coding Standards](http://mooseframework.org/wiki/CodeStandards/).
- Submit or improve [Test Cases](http://mooseframework.org/wiki/MooseTraining/testing/).
- Reference a specific issue, place "refs #<issue>" or "closes #<issue>" (e.g., #closes #1234).
-->

hugary1995 committed:
XFEM geometric cut switching material system

Add SwitchingMaterial, ComboCutUserObject, GeometricCutSubdomainIDAux and tests

close #17041

hugary1995 commented:
@bwspenc and @jiangwen84 this is part 2 of the refactoring.

[https://github.com/idaholab/moose/blob/2bee01da8fb7d88bb9e77e98d49cdaf8dbd601eb/modules/xfem/test/tests/switching_material/two_cuts_moving.i](https://github.com/idaholab/moose/blob/2bee01da8fb7d88bb9e77e98d49cdaf8dbd601eb/modules/xfem/test/tests/switching_material/two_cuts_moving.i) pretty much demonstrates all components of the new design:
```
[Problem]
  solve = false
[]

[XFEM]
  qrule = volfrac
  output_cut_plane = true
[]

[UserObjects]
  [cut1]
    type = LevelSetCutUserObject
    level_set_var = phi1
    negative_id = 1
    positive_id = 33
    execute_on = NONE
  []
  [cut2]
    type = LevelSetCutUserObject
    level_set_var = phi2
    negative_id = 5
    positive_id = 16
    execute_on = NONE
  []
  [combo]
    type = ComboCutUserObject
    cuts = 'cut1 cut2'
    keys = '1 5
            1 16
            33 5
            33 16'
    vals = '1 3 5 7'
    heal_always = true
  []
[]

[Mesh]
  [square]
    type = GeneratedMeshGenerator
    dim = 2
    nx = 5
    ny = 5
  []
[]

[AuxVariables]
  [phi1]
  []
  [phi2]
  []
  [cut1_id]
    order = CONSTANT
    family = MONOMIAL
  []
  [cut2_id]
    order = CONSTANT
    family = MONOMIAL
  []
  [combo_id]
    order = CONSTANT
    family = MONOMIAL
  []
[]

[AuxKernels]
  [phi1]
    type = FunctionAux
    variable = phi1
    function = 'x-0.213-t'
    execute_on = 'INITIAL TIMESTEP_BEGIN'
  []
  [phi2]
    type = FunctionAux
    variable = phi2
    function = 'x-0.728+t'
    execute_on = 'INITIAL TIMESTEP_BEGIN'
  []
  [cut1_id]
    type = GeometricCutSubdomainIDAux
    variable = cut1_id
    cut = cut1
  []
  [cut2_id]
    type = GeometricCutSubdomainIDAux
    variable = cut2_id
    cut = cut2
  []
  [combo_id]
    type = GeometricCutSubdomainIDAux
    variable = combo_id
    cut = combo
  []
[]

[Materials]
  [diffusivity_A]
    type = GenericConstantMaterial
    prop_names = 'A_D'
    prop_values = '5'
  []
  [diffusivity_B]
    type = GenericConstantMaterial
    prop_names = 'B_D'
    prop_values = '0.5'
  []
  [diffusivity_C]
    type = GenericConstantMaterial
    prop_names = 'C_D'
    prop_values = '12'
  []
  [diffusivity_D]
    type = GenericConstantMaterial
    prop_names = 'D_D'
    prop_values = '9'
  []
  [diff_combined]
    type = GeometricCutSwitchingMaterialReal
    base_name_keys = '1 3 5 7'
    base_name_vals = 'A B C D'
    prop_name = D
    geometric_cut_userobject = combo
    outputs = 'exodus'
    output_properties = 'D'
  []
[]

[Executioner]
  type = Transient

  dt = 0.25
  num_steps = 2

  max_xfem_update = 1
[]

[Outputs]
  exodus = true
[]
```
We can discuss how to fit the 3D cut (using a 2D mesh) into this new system.

moosebuild commented:
Job [Documentation](https://civet.inl.gov/job/685439/) on efbdf21 wanted to post the following:

View the site [here](https://mooseframework.inl.gov/docs/PRs/17042/site)

This comment will be updated on new commits.

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/688654/) on 5d5eb97 wanted to post the following:

Your code requires style changes.  

A patch was auto generated and copied [here](https://mooseframework.inl.gov/docs/PRs/17042/style.patch)  
You can directly apply the patch by running, in the top level of your repository:

`curl -s https://mooseframework.inl.gov/docs/PRs/17042/style.patch | git apply -v`  

Alternatively, with your repository up to date and in the top level of your repository:

`git clang-format 7b30e5ab9fa379e5a73bfc3c375e3fac56251206`

bwspenc commented on modules/xfem/doc/content/source/userobjects/ComboCutUserObject.md:
```diff
@@ -0,0 +1,17 @@
+
+## Overview
+
+The `ComboCutUserObject` combines multiple `GeometricCutUserObject`s to create a composite cut. Currently, all `GeometricCutUserObject`s are assumed to be non-overlapping, i.e. each element can only be cut by one geometric cut.
```
```
The `ComboCutUserObject` combines multiple instances of `GeometricCutUserObject` to define the geometric cut subdomain ID for all portions of the domain based on which side that point is located in relative to multiple cuts defined by `GeometricCutUserObject` instances. Currently, all `GeometricCutUserObject`s are assumed to be non-overlapping, i.e. each element can only be cut by one geometric cut.

bwspenc commented on modules/xfem/doc/content/source/userobjects/ComboCutUserObject.md:
```diff
@@ -0,0 +1,17 @@
+
+The `ComboCutUserObject` combines multiple `GeometricCutUserObject`s to create a composite cut. Currently, all `GeometricCutUserObject`s are assumed to be non-overlapping, i.e. each element can only be cut by one geometric cut.
+
+Each `GeometricCutUserObject` may have defined its own set of `GeometricCutSubdomainID`s, and the `ComboCutUserObject` follows the key-value dictionary provided by the user to assign new `GeometricCutSubdomainID` to each of the cut-subdomain.
```
to assign new -> to assign a

bwspenc replied:
cut-subdomain -> geometric cut subdomains.

bwspenc commented on modules/xfem/include/base/XFEM.h:
```diff
@@ -323,12 +334,17 @@ class XFEM : public XFEMInterface
-  std::map<const Elem *, const GeometricCutUserObject *> _healed_cuts;
+  typedef std::tuple<const Elem *, const GeometricCutUserObject *, GeometricCutSubdomainID>
+      GeometricCutElemInfo;
+  std::map<const Elem *, GeometricCutElemInfo> _geom_cut_elems;
```
Please add doxygen one-liners for each of these

bwspenc commented on modules/xfem/include/base/XFEM.h:
```diff
@@ -424,26 +440,19 @@ class XFEM : public XFEMInterface
-   * @param parent_elem The parent element
-   * @param cut_elem    The element being cut
-   * @param gcuo        The GeometricCutUserObject for the cut
+   * Randomly pick one node on the element that is in the physical domain
```
I'd reword this as:

Return the first node in the provided element that is found to be in the physical domain

...because it's not actually random.

bwspenc commented on modules/xfem/include/materials/GeometricCutSwitchingMaterial.h:
```diff
@@ -0,0 +1,82 @@
+
+/**
+ * switches between materials in a multi-material system where the
+ * interfaces are defined by multiple level set functions.
```
multiple geometric cut objects, right?

bwspenc commented on modules/xfem/include/materials/GeometricCutSwitchingMaterial.h:
```diff
@@ -0,0 +1,82 @@
+  GeometricCutSwitchingMaterialTempl(const InputParameters & parameters);
+
+protected:
+  // At the time of initiating this switching material, the initeQpStatefulProperties methods of the
```
initiating -> initializing

bwspenc reviewed (changes requested):
I haven't had a chance to go through all of this yet, and have to sign off for the day, but here are some relatively minor things to address. I think this is really close to being ready to go.

hugary1995 commented:
@bwspenc I fixed all that.

moosebuild commented:
Job [Parallel sweep evens](https://civet.inl.gov/job/709493/) on afd82e3 : invalidated by @hugary1995

failing test unrelated

hugary1995 commented:
@jiangwen84 @dschwen Could you please review this PR when you get a chance? There are some significant changes/additions to the XFEM module, so I think it's better to have more eyes on it.

The algorithmic flow is summarized here [https://mooseframework.inl.gov/docs/PRs/17042/site/modules/xfem/theory/embedded_interface.html](https://mooseframework.inl.gov/docs/PRs/17042/site/modules/xfem/theory/embedded_interface.html), and I tried my best to make the source code adequately commented.

bwspenc commented on modules/xfem/include/base/XFEM.h:
```diff
@@ -417,33 +474,59 @@ class XFEM : public XFEMInterface
-  GeometricCutSubdomainID getGeometricCutSubdomainID(const Elem * parent_elem,
-                                                     const Elem * cut_elem,
-                                                     const GeometricCutUserObject * gcuo) const;
+  const Node * pickOnePhysicalNode(const Elem * e, const Elem * e0) const;
```
It might be clearer to rename this to `pickFirstPhysicalNode`, but do as you see fit.

bwspenc commented on modules/xfem/src/materials/GeometricCutSwitchingMaterial.C:
```diff
@@ -0,0 +1,103 @@
+
+  // We may run into situations where the key doesn't exist in the base_name_map. This may happen
+  // when the problem is not well defined, or the level-set variables are very nonlinear, so that
+  // the combination of levelset signs are not provided. Cutting the timestep may help, so we throw
```
Would it make sense to query the geometric cut userobject for the list of keys it can provide at initialization and make sure we cover all of the possibilities then rather than during the computation?

hugary1995 replied:
Yeah that's the right thing to do. This check was carried over from my old design.

hugary1995 replied:
On a second thought, it's actually difficult to do the check at construction time because I can't query all the possible cut subdomain IDs for a given geometric cut userobject. Unless we require all geometric cut userobjects to explicit define their possible cut subdomain IDs.

bwspenc commented on modules/xfem/src/materials/GeometricCutSwitchingMaterial.C:
```diff
@@ -0,0 +1,103 @@
+  params.addRequiredParam<UserObjectName>("geometric_cut_userobject",
+                                          "The geometric cut userobject");
+  params.addRequiredParam<std::vector<GeometricCutSubdomainID>>(
+      "base_name_keys",
```
I think it might be clearer to call these parameters `geometric_cut_subdomain_ids` and `base_names`. We can discuss, though, if you don't like those suggestions.

hugary1995 replied:
`geometric_cut_subdomain_ids` and `base_names` are better. Again, these parameter names were carried over from the previous design where the key/value naming made more sense.

bwspenc commented on modules/xfem/src/userobjects/ComboCutUserObject.C:
```diff
@@ -0,0 +1,119 @@
+  params.addRequiredParam<std::vector<UserObjectName>>("cuts",
+                                                       "Vector of cut userobjects to combine");
+  params.addRequiredParam<std::vector<GeometricCutSubdomainID>>(
+      "keys", "Possible combinations of the geometric cut subdomain IDs");
```
Sorry, I'm not crazy about `keys` and `vals` either. Maybe `input_cut_ids` and `output_cut_ids`?

hugary1995 replied:
I want to somehow convey the message that we are doing some sort of look-up here. What are synonyms for "key-value"?

bwspenc commented on modules/xfem/src/userobjects/ComboCutUserObject.C:
```diff
@@ -0,0 +1,119 @@
+ComboCutUserObject::buildMap()
+{
+  if (_keys.size() % _num_cuts != 0)
+    mooseError("Expect multiples of ", _num_cuts, " keys, but got ", _keys.size(), " keys.");
```
'Expect' -> 'Expected'

bwspenc commented on modules/xfem/src/userobjects/ComboCutUserObject.C:
```diff
@@ -0,0 +1,119 @@
+  unsigned int num_combos = _keys.size() / _num_cuts;
+
+  if (_vals.size() != num_combos)
+    mooseError("Expect one value for each of the combination, got ",
```
'Expect'->'Expected'

bwspenc commented on modules/xfem/src/userobjects/ComboCutUserObject.C:
```diff
@@ -0,0 +1,119 @@
+  std::vector<GeometricCutSubdomainID> combo_key;
+  for (auto cut : _cuts)
+    combo_key.push_back(cut->getGeometricCutSubdomainID(node));
+  return _combo_ids.at(combo_key);
```
It seems like there's a chance that a particular combination might not exist, so we might need to handle that error. I'm saying that because I don't think the check on line 46 of this class exhaustively checks that you have every combination covered.

hugary1995 replied:
Yeah, that check in `GeometricCutSwitchingMaterial.C` should be moved here.

bwspenc reviewed (changes requested):
Overall, looks really good!  I'm excited to get this in. Just some really minor picky stuff.

hugary1995 committed:
address comments, fix sqa and pictures, clang

hugary1995 commented on modules/xfem/include/base/XFEM.h:
```diff
@@ -40,6 +40,52 @@ enum XFEM_QRULE
+ * (component 1) entry is a map for older material properties. The second entry will be empty if
+ * the material storage doesn't have older material properties.
+ */
+typedef std::array<std::unordered_map<unsigned int, std::string>, 2> CachedMaterialProperties;
```
If there is not stateful properties, the `_geom_cut_elems` storage will be empty. See `XFEM::storeMaterialPropertiesForElementHelper`.

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/724065/) on 74e3205 wanted to post the following:

Your code requires style changes.  

A patch was auto generated and copied [here](https://mooseframework.inl.gov/docs/PRs/17042/style.patch)  
You can directly apply the patch by running, in the top level of your repository:

`curl -s https://mooseframework.inl.gov/docs/PRs/17042/style.patch | git apply -v`  

Alternatively, with your repository up to date and in the top level of your repository:

`git clang-format 06bc1f73f0d4dfb3d01ce9ba77d6bb97346e07eb`

hugary1995 commented on modules/xfem/include/userobjects/ComboCutUserObject.h:
```diff
@@ -0,0 +1,109 @@
+
+  /**
+   * Loop over all the provided GeometricCutUserObjects, fill the data structures based on the first
+   * cut that wants to cut this 2D element.
```
Unfortunately it is (as of now). If multiple cuts want to cut the same element, how should we handle it?

bwspenc replied:
If we get a reasonable error in that case, that's fine for now. I assume that the cutter objects are doing the right thing. We just need to make sure we don't forget this when we add the ability for multiple cuts.

bwspenc commented on modules/xfem/doc/content/modules/xfem/theory/embedded_interface.md:
```diff
@@ -0,0 +1,50 @@
+
+## Moving interface
+
+To represent the movement of the interface, the user only need to update the interface location defined by the corresponding `GeometricCutUserObject`. The XFEM module handles the update of the interface location using the so-called *healing-and-re-cut* algorithm. In a nutshell, every time a interface location is updated, its corresponding cut (as well as the underlying strong discontinuity) is co ...
```
the user only need -> the user only needs

bwspenc commented on modules/xfem/doc/content/modules/xfem/theory/embedded_interface.md:
```diff
@@ -0,0 +1,50 @@
+       style=width:100%;padding:20px;
+       caption=The *healing-and-re-cut* algorithm at time step 1. For demonstration purposes, the `CEI`s are denoted using element IDs and the cut ID, while in the actual implementation pointers are used to store these information.
+
+During the first solve in the first time step, the mesh is not yet cut by XFEM, and the `CEI`s are empty. Therefore, nothing happens during function call `healMesh`. The cut marks element 1 for cutting, and it is cut into children elements 3 and 4 during the function call `cutMeshWithEFA`. When child element 3 is created, its parent element (1), the correponding cut (1), and its cut subdomain (1) ...
```
healMesh -> XFEM::healMesh()

bwspenc commented on modules/xfem/doc/content/modules/xfem/theory/embedded_interface.md:
```diff
@@ -0,0 +1,50 @@
+       style=width:100%;padding:20px;
+       caption=The *healing-and-re-cut* algorithm at time step 2. For demonstration purposes, the `CEI`s are denoted using element IDs and the cut ID, while in the actual implementation pointers are used to store these information.
+
+In the second time step, the system is first solved with updated boundary conditions, during which the signed distance function may be updated. In this case, the interface moves towards the right. During the function call `healMesh`, the cut is removed by deleting element 4. Since child element 3 is now treated as a potential parent element, the corresponding entries in `CEI` are updated to refle ...
```
healMesh->XFEM::healMesh()

bwspenc commented on modules/xfem/test/tests/switching_material/two_cuts_moving.i:
```diff
@@ -0,0 +1,135 @@
+  [combo]
+    type = ComboCutUserObject
+    geometric_cut_userobjects = 'cut1 cut2'
+    cut_subdomain_combinations = '1 5 1 16 33 5 33 16'
```
We should probably break all of these up into separate lines for clarity (in this and the other input files). I'm guessing that hit format doesn't play nicely with that, though.

hugary1995 replied:
This is now taking vector of vector of IDs per your suggestion. Input files are also updated.

bwspenc reviewed (changes requested):
Thanks for making those changes! I like the new syntax. Just a couple of really minor things that I saw in a second pass through. I'm not sure whether we're good to merge things just yet, but fix those up, and we'll have this all ready to go early next week when things are moving again.

hugary1995 committed:
address comments

hugary1995 committed:
Actually use TransientInterface!

hugary1995 commented:
Thanks Ben, I fixed all that. I also removed all the `Real time` arguments in all `GeometricCutUserObject`s -- it inherits `TransientInterface` so that seems redundant to me.

hugary1995 committed:
use vector of vector of CutSubdomainID

bwspenc reviewed (approved):
Looks good, thanks!
